# Handle Error

# Mục lục nội dung

- [1. Custom error](#1-custom-error)
- [2. Xử lý bọc error](#2-xử-lý-bọc-error)
- [3. Xử lý bọc router](#3-xử-lý-bọc-router)
- [4. Handle error](#4-handle-error)

## 1. Custom error

Lúc này ta có thể throw các error bên dưới trong service thay vì phải return về error

```js
// error.response.js
const StatusCode = {
  FORBIDEN: 403,
  CONFLICT: 409,
}
const ReasonStatusCode = {
  FORBIDEN: 'Bad request error',
  CONFLICT: 'Conflict error',
}
class ErrorResponse extends Error {
  constructor(message, status) {
    super(message)
    this.status = status
  }
}
class ConflictRequestError extends ErrorResponse {
  constructor(
    message = ReasonStatusCode.CONFLICT,
    statusCode = StatusCode.CONFLICT
  ) {
    super(message, statusCode)
  }
}
class BadRequestError extends ErrorResponse {
  constructor(
    message = ReasonStatusCode.FORBIDEN,
    statusCode = StatusCode.FORBIDEN
  ) {
    super(message, statusCode)
  }
}
module.exports = {
  ConflictRequestError,
  BadRequestError,
}
```

**[⬆ Quay trở lại đầu trang](#mục-lục-nội-dung)**

## 2. Xử lý bọc error

- Xử lý bọc router và middleware hạn chế sử dụng try catch

```js
// asyncHandler.js
const asyncHandler = (fn) => {
  return (req, res, next) => {
    fn(req, res, next).catch(next)
  }
}
module.exports = asyncHandler
```

**[⬆ Quay trở lại đầu trang](#mục-lục-nội-dung)**

## 3. Xử lý bọc router

Ban đầu thì service sẽ try catch và ném lỗi về cho controller và controller sẽ try catch và ném error về app xử lý thì khi ta bọc router như dưới thì ta sẽ bỏ hết try catch ở service và controller

```js
// index - access
const { asyncHandler } = require('../../auth/checkAuth')
const router = express.Router()

// service của controller này không cần try catch
router.post('/shop/signup', asyncHandler(accessController.signup))

module.exports = router
```

**[⬆ Quay trở lại đầu trang](#mục-lục-nội-dung)**

## 4. Handle error

Xử lý bắt error ném về

```js
// app.js
...
// init routes------------------------------------------------------------------
app.use('/', require('./routers'))

// handling error---------------------------------------------------------------
app.use((req, res, next) => { // tạo error 404 xử lý khi router call về không đúng
  const error = new Error('Not found')
  error.status = 404
  next(error) // chuyển sang middleware xử lý error
})

app.use((error, req, res, next) => { // xử lý đọc error (bao gồm error ở trên)
  const statusCode = error.status || 500
  return res.status(statusCode).json({
    status: 'error',
    code: statusCode,
    message: error.message || 'Internal Server Error',
  })
})
```

**[⬆ Quay trở lại đầu trang](#mục-lục-nội-dung)**

## Code tham khảo

- Xem code đầy đủ [ở đây](https://github.com/thonghp/ecommerce-node-js/commit/4c1e2020bcaa46e9cef5e77c806829475fb88dcc)
- [Trang chủ](../README.md)
